### PR TITLE
Adding ability to delete by selector.  Also variable name cleanup and update to use .format.

### PR DESCRIPTION
**Pull request by kwoodson**

### CODE REVIEW

**Comment by kwoodson** [comment]
The major changes here:

- `lib/base.py`:  These changes allow specifically `oc_obj` to delete by name or by selector.  They propagated to all files as a consequence.  PR looks large but change set is small.
- I did not like `rname` as it used to conflict with the way ansible `import *` was required which conflicted with the name variable.  I renamed back to `name` as we no longer do the import the same way.
- I added an `integration/oc_obj.yml` test file.  It failed first, then it passed when my changes were added.
- I removed the restriction set on deletion requiring a `name` being passed in `class/oc_obj.py`.

File: roles/lib_openshift/src/lib/base.py [hunk 1]
```diff
@@ -95,11 +95,15 @@ def _create(self, fname):
         '''call oc create on a filename'''
         return self.openshift_cmd(['create', '-f', fname])
 
-    def _delete(self, resource, rname, selector=None):
+    def _delete(self, resource, name=None, selector=None):
         '''call oc delete on a resource'''
         cmd = ['delete', resource, rname]
```
Comment (jarrpa) [inline roles/lib_openshift/src/lib/base.py]: Forgot to remove `rname`.

File: roles/lib_openshift/library/oc_adm_ca_server_cert.py [hunk 2]
```diff
@@ -919,11 +919,15 @@ def _create(self, fname):
         '''call oc create on a filename'''
         return self.openshift_cmd(['create', '-f', fname])
 
-    def _delete(self, resource, rname, selector=None):
+    def _delete(self, resource, name=None, selector=None):
         '''call oc delete on a resource'''
         cmd = ['delete', resource, rname]
-        if selector:
-            cmd.append('--selector=%s' % selector)
+        if selector is not None:
```
Comment (ashcrow) [inline roles/lib_openshift/library/oc_adm_ca_server_cert.py]: 👍

File: roles/lib_openshift/library/oc_adm_ca_server_cert.py [hunk 3]
```diff
@@ -941,7 +945,7 @@ def _process(self, template_name, create=False, params=None, template_data=None)
         else:
             cmd.append(template_name)
         if params:
-            param_str = ["%s=%s" % (key, value) for key, value in params.items()]
+            param_str = ["{}={}".format(key, value) for key, value in params.items()]
```
Comment (ashcrow) [inline roles/lib_openshift/library/oc_adm_ca_server_cert.py]: 👍

File: roles/lib_openshift/library/oc_adm_ca_server_cert.py [hunk 4]
```diff
@@ -958,13 +962,13 @@ def _process(self, template_name, create=False, params=None, template_data=None)
...
 
-    def _get(self, resource, rname=None, selector=None):
+    def _get(self, resource, name=None, selector=None):
         '''return a resource by name '''
         cmd = ['get', resource]
-        if selector:
-            cmd.append('--selector=%s' % selector)
-        elif rname:
-            cmd.append(rname)
+        if selector is not None:
```
Comment (ashcrow) [inline roles/lib_openshift/library/oc_adm_ca_server_cert.py]: Is this a "one or the other" thing as well? If so, should it raise if *both* are provided? Same question for other similar operations.

Comment (kwoodson) [inline roles/lib_openshift/library/oc_adm_ca_server_cert.py]: @ashcrow, good question.  I just tried a query with both given and it is an error when both are supplied.

```
oc get template mysql --selector=template=mysql-ephemeral-template --all-namespaces
error: name cannot be provided when a selector is specified
```

I will update the module params to be mutually-exclusive.

**Comment by kwoodson** [comment]
@jarrpa, good catch.  I pushed an update with rname removed.

**Comment by jarrpa** [comment]
@kwoodson It would be good to replicate some of the changes being made to `_delete()` in `_get()`, though that's a different PR. :)

File: roles/lib_openshift/library/oc_obj.py [hunk 5]
```diff
@@ -912,11 +912,15 @@ def _create(self, fname):
...
-        cmd = ['delete', resource, rname]
-        if selector:
-            cmd.append('--selector=%s' % selector)
+        cmd = ['delete', resource]
+        if selector is not None:
+            cmd.append('--selector={}'.format(selector))
+        elif name is not None:
+            cmd.append(name)
+        else:
+            raise OpenShiftCLIError('Either name or selector is required when calling delete.')
```
Comment (ewolinetz) [inline roles/lib_openshift/library/oc_obj.py]: 👍

**Comment by ashcrow** [comment]
```
************ Module oc_obj
roles/lib_openshift/library/oc_obj.py:1453: [C0330(bad-continuation), ] Wrong continued indentation before block.
           '\"%s\" not found' % self.name in results['stderr']):
           ^|   |
roles/lib_openshift/library/oc_obj.py:1650: [C0326(bad-whitespace), ] Exactly one space required after comma
        mutually_exclusive=[["content", "files"],["selector", "name"]],
```

**Comment by kwoodson** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
@ewolinetz @ashcrow, I think this is ready now.

**Comment by ashcrow** [comment]
aos-ci-test

**Comment by ashcrow** [comment]
[test]

File: roles/lib_openshift/src/test/integration/oc_obj.yml [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                - mysql
+                from:
+                  kind: ImageStreamTag
+                  name: mysql:latest
+              type: ImageChange
+
+  - name: delete using selector
+    oc_obj:
+      namespace: test
+      selector: name=mysql
```
Comment (ewolinetz) [inline roles/lib_openshift/src/test/integration/oc_obj.yml]: does this work as a selector? I would expect that the DC you specify above would have a value here:
https://github.com/openshift/openshift-ansible/pull/3840/files#diff-a1c29b1eb7b762844f88ac765306d442R138

Can we add a `list` operation right before this delete to verify that we see the object before we delete it as we do on 194?

**Comment by ewolinetz** [comment]
@kwoodson question regarding oc_obj test otherwise LGTM

**Comment by ashcrow** [comment]
Looks like another rebase has to occur.

**Comment by kwoodson** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
@ashcrow, yeah, a little tricky based on the `# pragma: no cover` that was merged before this.  I think its ready this time.

**Comment by ewolinetz** [comment]
@kwoodson i have a question regarding the `oc_obj` test

**Comment by kwoodson** [comment]
@ewolinetz, Your question regarding the oc_obj test was good.  I investigated it yesterday and found that when I run that oc_obj with `state: present` that when the `deploymentconfig` was created that it automatically added the `selector: name=mysql`.

```yaml
  spec:
    replicas: 1
    selector:
      name: mysql
```

**Comment by kwoodson** [comment]
Error in the CI tests.
```
OSError: [Errno 2] No such file or directory: '/tmp/pip-YCOWQf-record'
```

**Comment by kwoodson** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
@sdodson @stevekuznetsov 
```
GH_DESC="Internal Jenkins error during provisioning an 3.5 cluster"
```

**Comment by stevekuznetsov** [comment]
Is that in the AOS CI tests? I am not familiar with those at all unfortunately.

**Comment by kwoodson** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
[merge]

**Comment by kwoodson** [comment]
@ewolinetz @ashcrow would either of you mind :+1:  on this?

**Comment by stevekuznetsov** [comment]
@kwoodson you ran the tests a number of times over and over here without changing code -- what was the issue? If there are flakes in the tests, we should all be making issues and tracking the flakes so we can adequately prioritize fixing high-frequency flakes.

**Comment by kwoodson** [comment]
@stevekuznetsov, the tests were failing randomly and intermittently.  
- `No such file or directory`
- ```error: aos-ci-jenkins/OS_3.5_containerized for 92f3067 (logs)```

I'm not sure why but the tests are the worst part about getting PRs into this repo.

**Comment by stevekuznetsov** [comment]
@kwoodson a test failing intermittently is the textbook definition of a flake. Please log issues for them and help the organization work on eliminating them.

**Comment by kwoodson** [comment]
[merge]

**Comment by sdodson** [comment]
comments above say unit tests were fine, but the status in github is marked failed. I really hate to do this, but lets give this one more shot.

**Comment by sdodson** [comment]
aos-ci-test

**Comment by stevekuznetsov** [comment]
@arilivigni it looks like this PR status for `aos-ci-jenkins/OS_unit_tests` can't recover from having failed once before

**Comment by sdodson** [comment]
I don't actually see the unit tests having failed at any time in the past week. I made a change to add system container jobs, i'm going to revert that.

**Comment by sdodson** [comment]
aos-ci-test

**Comment by kwoodson** [comment]
flake https://github.com/openshift/origin/issues/10162

**Comment by kwoodson** [comment]
[merge]

**Comment by sdodson** [comment]
I just pushed the context smashing fix, so that merge is going to fail, aborting it.

**Pull request merged**